File: sdk/py/splendor_actor/types/player.py
from typing import List, Dict, Optional

from .card import CardView, DevelopmentCards, Card
from .color import ColorVec, Color
from .noble import Noble


class Player:
    """
    represent the player
    """

    idx: int
    points: int
    tokens: ColorVec
    development_cards: DevelopmentCards
    reserved_cards: List[CardView]
    nobles: List[Noble]

    def __init__(
        self,
        idx: int,
        points: int,
        tokens: List[int],
        development_cards: Dict,
        reserved_cards: List[Dict],
        nobles: List[Dict],
    ):
        self.idx = idx
        self.points = points
        self.tokens = ColorVec(tokens)
        self.development_cards = DevelopmentCards(**development_cards)
        self.reserved_cards = [CardView(**card) for card in reserved_cards]
        self.nobles = [Noble(**noble) for noble in nobles]

    def can_buy(self, card: Card) -> Optional[ColorVec]:
        """
        check if the player can buy the card, return the tokens needed if possible
        """
        effective_cost = card.requires.saturating_sub(self.development_cards.bonus)
        diff = effective_cost.saturating_sub(self.tokens)
        if diff.total() > self.tokens[Color.YELLOW]:
            return None
        uses = effective_cost - diff
        uses[Color.YELLOW] = diff.total()
        return uses

File: sdk/go/types/player.go
package types

type Player struct {
	Idx              uint8            `json:"idx"`
	Points           uint8            `json:"points"`
	Tokens           ColorVec         `json:"tokens"`
	DevelopmentCards DevelopmentCards `json:"development_cards"`
	ReservedCards    []CardView       `json:"reserved_cards"`
	Nobles           []Noble          `json:"nobles"`
}

// check if the player can buy the card, return the tokens needed if possible
func (p *Player) CanBuy(card *Card) (bool, *ColorVec) {
	effective_cost := card.Requires
	effective_cost.SaturatingSub(&p.DevelopmentCards.Bonus)
	diff := effective_cost
	diff.SaturatingSub(&p.Tokens)
	if diff.Total() > p.Tokens[YELLOW] {
		return false, nil
	}
	effective_cost.Sub(&diff)
	effective_cost[YELLOW] = diff.Total()
	return true, &effective_cost
}
